lib/helpers/signed-aws-es-fetch.js

```diff
@@ -0,0 +1,124 @@
+const aws4 = require('aws4');
+const nodeFetch = require('node-fetch');
+const logger = require('@financial-times/n-logger').default;
+const resolveCname = require('util').promisify(require('dns').resolveCname);
+const memoize = require('lodash/memoize');
+
+const memoizedResolveCname = memoize(resolveCname);
+
+const defaultAwsKeys = {
+	awsKeys: ['ES_AWS_ACCESS_KEY', 'AWS_ACCESS_KEY', 'AWS_ACCESS_KEY_ID'],
+	awsSecretKeys: ['ES_AWS_SECRET_ACCESS_KEY', 'AWS_SECRET_ACCESS_KEY'],
+	tokenKeys: ['ES_AWS_SESSION_TOKEN', 'AWS_SESSION_TOKEN']
+};
+
+module.exports = async function (url, opts, creds) {
+	opts = opts || {};
+	url = await resolveUrlAndHost(url);
+	return signedFetch(url, opts, creds);
+};
+
+function getURLObject(url) {
+	try {
+		const urlObject = new URL(url);
+		return urlObject;
+	} catch (error) {
+		logger.error({
+			event: 'Invalid URL',
+			data: {
+				url
+			}
+		});
+		throw error;
+	}
+}
+
+async function resolveValidHost(urlObject) {
+	try {
+		const hosts = await memoizedResolveCname(urlObject.host);
+		if (hosts && hosts[0]) {
+			return hosts[0];
+		}
+	} catch (error) {}
+	logger.error({
+		event: 'Invalid Host',
+		data: {
+			url: urlObject
+		}
+	});
+	throw new Error('Invalid Host');
+}
+
+function isActiveDnsResolution() {
+	return (
+		!process.env.AWS_SIGNED_FETCH_DISABLE_DNS_RESOLUTION ||
+		process.env.AWS_SIGNED_FETCH_DISABLE_DNS_RESOLUTION === 'false'
+	);
+}
+
+async function resolveUrlAndHost(url) {
+	const urlObject = getURLObject(url);
+	if (
+		!/\.es\.amazonaws\.com$/.test(urlObject.host) &&
+		isActiveDnsResolution()
+	) {
+		const host = await resolveValidHost(urlObject);
+		url = url.replace(urlObject.host, host);
+	}
+	return url;
+}
+
+function defaultAwsCredentials(keys) {
+	const keyName = defaultAwsKeys[keys].find((keyName) => process.env[keyName]);
+	return process.env[keyName];
+}
+
+function thereIsToken() {
+	return (
+		process.env.ES_AWS_SESSION_TOKEN !== false &&
+		// a boolean value is interpreted as string if set from vault
```

Comment:
This comment I think is a bit confusing - my understanding is that `ES_AWS_SESSION_TOKEN` can only be part of some [temporary credentials](https://docs.aws.amazon.com/STS/latest/APIReference/API_GetSessionToken.html) so it's impossible for it to come from Vault. It could come from the environment though.

Reply:
Yeah it's misleading because technically _every_ item in `process.env` is a string by default, regardless of where it comes from. The line above where we check explicitly for `false` could only happen if `process.env.ES_AWS_SESSION_TOKEN` is manually changed somewhere in the app or one of the dependencies.